README.md
# Math Intervals
[![Build Status](https://travis-ci.org/danielfdsilva/mathInterval.svg?branch=master)](https://travis-ci.org/danielfdsilva/mathInterval)  

This library lets you use mathematical intervals in php. Trust me, it's way cooler than it sounds :)

Before using this library you should understand what intervals are and how to use them.
[MathIsFun](http://www.mathsisfun.com/sets/intervals.html) does an excellent job at explaining them so take a look.

### Interval Notation
**MathInterval** uses square brackets to define an interval. The direction of each bracket indicates whether we want to include or exclude the beginning and ending number. For example:

```]5,12] - From 5 (excluded) to 12 (included)```  
and  
```]5,12[ - From 5 (excluded) to 12 (excluded)```

This notation is accepted as part of  [ISO 31-11](http://en.wikipedia.org/wiki/ISO_31-11).

**Important note:** For the time being there's no way to define an interval to infinity.

## Using MathInterval

To create a MathInterval instance simply use:
```php
$interval = new MathInterval('[5,20]');
```
With this you create an interval that comprehends all numbers from 5 to 20 (including 5 and 20).  
Now you can start validating number against this interval:
```php
$interval->inInterval(5); // TRUE
$interval->inInterval(10); // TRUE
$interval->inInterval(150); // FALSE

$interval->inInterval(15.25); // FALSE
```
What?! How come 15.25 is ```FALSE```. Isn't it between 5 and 20?  

To simplify validation, MathInterval assumes that only integer values are valid, but don't despair. If you need to allow float numbers you just need to initialize the interval with one of the values as float:

```php
// Initializing one of the values as float, makes the interval handle them.
$interval = new MathInterval('[5.0,20]');

$interval->inInterval(15.25); // TRUE
```

### Union

```php
// Create your first interval.
$interval = new MathInterval('[5,20]');

// Unite with another interval.
$interval->union('[17,33[');

// The resulting interval will be, of course, the union of both.
print $interval; // [5,33[
```

### Intersection
The intersection of intervals works in the same way as union, but with a very different outcome.
```php
// Create your first interval.
$interval = new MathInterval('[5,20]');

// Intersect with another interval.
$interval->intersection('[17,33[');

// The resulting interval will be, of course, the intersection of both.
print $interval; // [17,20]
```

### Expressions
MathInterval also allows you to use expressions to initialize and compute intervals. The official symbols for intersections and union are ∩ and ⋃, but, since these are hard to type, MathInterval uses ```and``` and ```or```.

For example:


```php
// [5,20] ⋃ [10,25]
$interval = new MathIntervalCollection('[5,20] or [10,25]');
print $interval; // [5,25]

// [5,20] ∩ [10,25]
$interval = new MathIntervalCollection('[5,20] and [10,25]');
print $interval; // [10,20]

// You can chain as many values as needed in an expression:
// [5,20] ∩ [10,25] ⋃ [15,30[
$interval = new MathIntervalCollection('[5,20] and [10,25] or [15,30[');
print $interval; // [10,30[

// You can also use these expressions in the union() and intersection() methods.
```
Just like in any mathematical expression, the order of the operators matters and parenthesis can be used to specify the order of the operations.  
For example: ```[5,20] ∩ ([10,25] ⋃ [15,30[)``` = ```[5,20] ∩ [10,30[``` = ```[10,20]```

**With complex expressions you must use `MathIntervalCollection`. All the methods available in `MathInterval` like union and intersection are also available here.**

```php
$interval = new MathIntervalCollection('[1,10]');
$interval->union('[16,20]');
// Note that some intervals are not possible to unite therefore they
// will be left as an expression.
print $interval; // [1,10] or [16,20]


$interval = new MathIntervalCollection('[1,10] or [16,20]');
$interval->intersection(']7,9] or [15,18]');
print $interval; // ]7,9] or [16,18]
// This would be the same as ([1,10] or [16,20]) and (]7,9] or [15,18])
```

-----

## Contribution
You are free to contribute to the project. If you find a bug or have a nice idea about a feature, please open an issue or submit your own solution. I'll be more than happy to hear your suggestions. :)

## Testing
The framework testing is done using phpunit. To run the tests you just need to run ```phpunit``` in the root folder.

##License
MathInterval is licensed under **The MIT License (MIT)**, see the LICENSE file for more details.


changelog.md
## Changelog

### 2.0.0

- Union and intersection methods accept MathInterval objects. ~~Fix #1~~
- Add MathIntervalCollection to support correct union of intervals. ~~Fix #4~~

### 1.1.0

- Add support for expressions.

### 1.0.0

- Initial release.

tests/bootstrap.php
<?php

require "src/MathIntervalCollection.php";


src/MathRange.php
<?php

// In class constants you can't define constants with concatenation
// because of reasons.
// Match: [1,2].
define('MATH_RANGE_REGEX', '(\[|\])(-?[0-9]+(?:\.?[0-9]+)?),(-?[0-9]+(?:\.?[0-9]+)?)(\[|\])');
// Match: [1,2] or [3,4] and [5,6].
define('MATH_RANGE_EXPRESSION_REGEX', MATH_RANGE_REGEX . '(?: (?:or|and) '.MATH_RANGE_REGEX.')*');
// Match: ([1,2] or [3,4]).
define('MATH_RANGE_EXP_ATOM_REGEX', '\(('.MATH_RANGE_EXPRESSION_REGEX.'\)');

class MathRange {

  private $lBoundIn = NULL;
  private $lBound = NULL;
  private $uBound = NULL;
  private $uBoundIn = NULL;
  private $allowFloat = TRUE;
  private $emptyRange = FALSE;
  
  function __construct($expression) {
    $range = MathRange::compute($expression);

    // Check for empty.
    switch ($range) {
      case ']0,0[':
         $this->allowFloat = FALSE;
      case ']0.0,0.0[':
        // If compute returned an empty range initialize as such.
        return $this->setEmpty();
      break;
    }

    // Chop pieces. Validation was already done in compute().
    preg_match('/^'.MATH_RANGE_REGEX.'$/', $range, $pieces);
    list(, $lbound_in_ex, $lbound, $ubound, $ubound_in_ex) = $pieces;
    // Convert upper and lower bound to number by adding 0.
    $lbound += 0;
    $ubound += 0;

    $this->lBoundIn = $lbound_in_ex == '[' ? TRUE : FALSE;
    $this->lBound = $lbound;
    $this->uBound = $ubound;
    $this->uBoundIn = $ubound_in_ex == ']' ? TRUE : FALSE;

    $this->allowFloat = !(is_int($lbound) && is_int($ubound));
  }

  function __toString() {
    if ($this->isEmpty()) {
      if ($this->allowFloats()) {
        return ']0.0,0.0[';
      }
      else {
        return ']0,0[';
      }
    }
    
    $out = '';

    $out .= $this->includeLowerBound() ? '[' : ']';
    // If the bounds is something like 1.0, the 0 won't be printed.
    // To correctly allow decimals we need it.
    $out .= $this->getLowerBound();    
    if ($this->allowFloats()) {
      $decimal_part = $this->getLowerBound() - floor($this->getLowerBound());
      if ($decimal_part == 0) {
        $out .= '.0';
      }
    }

    $out .= ',';

    // If the bounds is something like 1.0, the 0 won't be printed.
    // To correctly allow decimals we need it.
    $out .= $this->getUpperBound();    
    if ($this->allowFloats()) {
      $decimal_part = $this->getUpperBound() - floor($this->getUpperBound());
      if ($decimal_part == 0) {
        $out .= '.0';
      }
    }

    $out .= $this->includeUpperBound() ? ']' : '[';
    
    return $out;
  }

  static function compute($expression) {
    if (preg_match('/^'.MATH_RANGE_REGEX.'$/', $expression, $pieces)) {
      // Extracted range.
      list(, $lbound_in_ex, $lbound, $ubound, $ubound_in_ex) = $pieces;
      // Convert upper and lower bound to number by adding 0.
      $lbound += 0;
      $ubound += 0;

      // Validate range.
      if ($lbound > $ubound) {
        throw new MathRangeException("Lower bound must be lower than upper bound in $expression");
      }
      elseif (($ubound > $lbound) || ($lbound == $ubound && $lbound_in_ex == '[' && $ubound_in_ex == ']')) { 
        // [1,1] is a valid range allowing only the number 1.
        return $expression;
      }
      else {
        // ]1,1], [1,1[ and ]1,1[ are all empty.
        // Return as empty range .]0,0[ Accounting for floats.
        return (is_int($lbound) && is_int($ubound)) ? ']0,0[' : ']0.0,0.0[';
      }
    }
    /*elseif (preg_match('/^'.MATH_RANGE_EXPRESSION_REGEX.'$/', $expression, $results)) {
      // There are no atoms to extract.
      // Proceed to compute.
      $atom = $results[0];
      $ranges = preg_split('/ /', $atom);
      
      // First element is always a range.
      // The regex already validated this.
      $working_range = new MathRange(array_shift($ranges));
      $op = current($ranges);
      
      do {
        $exp = next($ranges);
        
        if ($op == 'or') {
          $working_range->union($exp);
        }
        elseif ($op == 'and') {
          $working_range->intersection($exp);
        } 
      }
      while ($op = next($ranges));
      
      return $working_range->__toString();
    }*/
    /*elseif (preg_match('/'.MathRange::RANGE_EXP_ATOM_REGEX.'/', $expression, $atoms)) {
      // Extracted atom.
      $atom = $atoms[1];
      print "$atom\n";
      compute($atom);
    }*/
    else {
      throw new MathRangeException("Invalid expression.");
    }
  }

  private function setEmpty() {
    $this->lBoundIn = FALSE;
    $this->lBound = 0;
    $this->uBound = 0;
    $this->uBoundIn = FALSE;
    $this->emptyRange = TRUE;
    
    return $this;
  }
  
  public function getUpperBound() {
    return $this->uBound;
  }
  
  public function getLowerBound() {
    return $this->lBound;
  }
  
  public function includeLowerBound() {
    return $this->lBoundIn;
  }
  
  public function includeUpperBound() {
    return $this->uBoundIn;
  }
  
  public function isEmpty() {
    return $this->emptyRange;
  }
  
  public function allowFloats() {
    return $this->allowFloat;
  }
  
  public function union($expression) {
    $toJoin = new MathRange($expression);
    
    // Handle float values.
    $this->allowFloat = $this->allowFloat || $toJoin->allowFloats();
    
    // Handle empty ranges.
    if ($toJoin->isEmpty()) {
      return $this;
    }
    elseif ($this->isEmpty()) {
      // Copy $toJoin.
      $this->lBoundIn = $toJoin->includeLowerBound();
      $this->lBound = $toJoin->getLowerBound();
      $this->uBound = $toJoin->getUpperBound();
      $this->uBoundIn = $toJoin->includeUpperBound();
      // Allowing floats depends on both ranges. Do not copy this property.
      // $this->allowFloat = $toJoin->allowFloats();
      $this->emptyRange = $toJoin->isEmpty();
      return $this;
    }
    
    // No empty ranges. Unite.
    // Upper bound.
    if ($toJoin->getUpperBound() > $this->getUpperBound()) {
      // Since the upper bound of the expression to join is higher
      // the new upper bound is the one from that expression.
      $this->uBound = $toJoin->getUpperBound();
      $this->uBoundIn = $toJoin->includeUpperBound();
    }
    elseif ($this->getUpperBound() == $toJoin->getUpperBound()) {
      // With equal upper bounds, if one is included the resulting
      // union also has an included upper bound.
      if ($toJoin->includeUpperBound()) {
        $this->uBoundIn = TRUE;
      }
    }
    // else the current rage stays as is.
    
    // Lower bound.
    if ($toJoin->getLowerBound() < $this->getLowerBound()) {
      // Since the lower bound of the expression to join is lower
      // the new lower bound is the one from that expression.
      $this->lBound = $toJoin->getLowerBound();
      $this->lBoundIn = $toJoin->includeLowerBound();
    }
    elseif ($this->getLowerBound() == $toJoin->getLowerBound()) {
      // With equal lower bounds, if one is included the resulting
      // union also has an included lower bound.
      if ($toJoin->includeLowerBound()) {
        $this->lBoundIn = TRUE;
      }
    }
    // else the current rage stays as is.
    
    return $this;
  }

  public function intersection($expression) {
    $toJoin = new MathRange($expression);
    
    // Handle float values.
    $this->allowFloat = $this->allowFloat || $toJoin->allowFloats();

    // Handle empty ranges.
    // Intersection with an empty range is always empty.
    if ($toJoin->isEmpty()) {
      return $this->setEmpty();
    }
    elseif ($this->isEmpty()) {
      return $this;
    }

    // No empty ranges. Intersect.
    // Check if $this contains $toJoin;
    if ($this->getLowerBound() < $toJoin->getLowerBound() && $this->getUpperBound() > $toJoin->getUpperBound()) {
      // Copy $toJoin.
      $this->lBoundIn = $toJoin->includeLowerBound();
      $this->lBound = $toJoin->getLowerBound();
      $this->uBound = $toJoin->getUpperBound();
      $this->uBoundIn = $toJoin->includeUpperBound();
      // Allowing floats depends on both ranges. Do not copy this property.
      // $this->allowFloat = $toJoin->allowFloats();
      $this->emptyRange = $toJoin->isEmpty();
      return $this;
    }
    // Check if $toJoin contains $this;
    elseif ($this->getLowerBound() > $toJoin->getLowerBound() && $this->getUpperBound() < $toJoin->getUpperBound()) {
      return $this;
    }
    
    // Find out which range comes first.
    if ($this->getLowerBound() <= $toJoin->getLowerBound() && $this->getUpperBound() <= $toJoin->getUpperBound()) {
      $first = $this;
      $second = $toJoin;
    }
    else {
      $first = $toJoin;
      $second = $this;
    }
    
    // Case:
    //  __________  _________
    //           | |
    //------------------------
    if ($first->getUpperBound() < $second->getLowerBound()) {
      // No intersection.
      return $this->setEmpty();
    }
    // Case: (the upper and lower are the same.)
    //  ___________________
    //           |
    //------------------------
    elseif ($first->getUpperBound() == $second->getLowerBound()) {
      // The only possible option here is a range with only one value allowed
      // like [1,1] but for that both bounds need to be included.
      if ($first->includeUpperBound() && $second->includeLowerBound()) {
        $this->lBoundIn = TRUE;
        $this->lBound = $first->getUpperBound();
        $this->uBound = $first->getUpperBound();
        $this->uBoundIn = TRUE;
        return $this;
      }
      else {
        // No intersection.
        return $this->setEmpty();
      }
    }
    // Case: (Ranges overlap. Account for inclusions.)
    //   _____________
    //  |____________|         --> $second
    //  |            |         --> $first
    //------------------------
    elseif ($first->getLowerBound() == $second->getLowerBound() && $first->getUpperBound() == $second->getUpperBound()) {
      // Values are the same. Set only inclusions.
      $this->lBoundIn = $first->includeLowerBound() && $second->includeLowerBound();
      $this->uBoundIn = $first->includeUpperBound() && $second->includeUpperBound();
      return $this;
    }
    // Case: (Ranges overlap on lower.)
    //   __________________
    //  |_____________    |    --> $second
    //  |            |         --> $first
    //------------------------
    elseif ($first->getLowerBound() == $second->getLowerBound() && $first->getUpperBound() < $second->getUpperBound()) {
      // Values are the same. Set only inclusions.
      $this->lBoundIn = $first->includeLowerBound() && $second->includeLowerBound();
      $this->lBound = $first->getLowerBound();
      $this->uBound = $first->getUpperBound();
      $this->uBoundIn = $first->includeUpperBound();
      return $this;
    }
    // Case: (Ranges overlap on upper.)
    //   __________________
    //  |    _____________|    --> $first
    //      |            |     --> $second
    //------------------------
    elseif ($first->getLowerBound() < $second->getLowerBound() && $first->getUpperBound() == $second->getUpperBound()) {
      // Values are the same. Set only inclusions.
      $this->lBoundIn = $second->includeLowerBound();
      $this->lBound = $second->getLowerBound();
      $this->uBound = $second->getUpperBound();
      $this->uBoundIn = $first->includeUpperBound() && $second->includeUpperBound();
      return $this;
    }
    // Having excluded all other cases the remaining one is:
    //         ____________
    //  ______|_______    |    --> $second
    //               |         --> $first
    //------------------------
    // The upper bound belongs to $first and
    // the lower bound belongs to $second.
    else {
      $this->lBoundIn = $second->includeLowerBound();
      $this->lBound = $second->getLowerBound();
      $this->uBound = $first->getUpperBound();
      $this->uBoundIn = $first->includeUpperBound();
      return $this;
    }
  }
}

class MathRangeException extends Exception { }

src/MathIntervalCollection.php
<?php
require 'MathInterval.php';

/**
 * @file MathIntervalCollection.php
 * 
 * @author Daniel da Silva
 * @package MathInterval
 * @version 2.0.0
 */

// In class constants you can't define constants with concatenation
// because of reasons.
// Match: [1,2] or [3,4] and [5,6].
define('MATH_INTERVAL_EXPRESSION_REGEX', MATH_INTERVAL_REGEX. '(?: (?:or|and) '.MATH_INTERVAL_REGEX.')*');
// Match: ([1,2] or [3,4]).
define('MATH_INTERVAL_EXP_ATOM_REGEX', '\(('.MATH_INTERVAL_EXPRESSION_REGEX.')\)');

/**
 * Library to use Mathematical Intervals.
 * 
 * Allows the creation of an interval and the validation
 * of values against said interval.
 * It is also possible to perform the union and intersection of
 * different intervals.
 */
class MathIntervalCollection implements Iterator {

  /**
   * MathInterval version.
   */
   const VERSION = MathInterval::VERSION;

   /**
   * Position of the intervals array.
   * Used by the iterator.
   * @var int
   * @access private
   */
   private $position = 0;

  /**
   * The array containing all the intervals belonging to the collection.
   * I.E. All the intervals to which a value can belong.
   * @var array
   * @access private
   */
   private $intervals = array();

   /**
    * The original expression with which the collection was initialized.
    */
   private $originalExpr = NULL;

  /**
   * MathIntervalCollection constructor.
   * Computes a collection of mathematical intervals from the provided
   * expression.
   * 
   * @uses MathIntervalCollection::compute($expression)
   * 
   * @param string $expression
   *   The interval to compute.
   * 
   * @throws MathIntervalException
   *   - If the provided expression is invalid.
   *   - If the lower bound of the interval is higher then its upper bound.
   * 
   * @return MathInterval
   */
  function __construct($expression) {
    $this->originalExpr = $expression;
    $expression = MathIntervalCollection::compute($expression);
    
    // The expression is already simplified. 
    // It will be in the form of [1,2] or [3,4]
    $ranges = preg_split('/ /', $expression);
    $tot = count($ranges);
    for ($i=0; $i < $tot; $i+=2) {
      $this->intervals[] = new MathInterval($ranges[$i]);
    }
  }

  /**
   * Provides a string representation of the interval that can be used
   * to initialize a new MathInterval.
   * 
   * @return string
   *   The string representation of the interval.
   */
  function __toString() {
    $outStrings = array_map(function($interval) {
        return $interval->__toString();
      }, $this->intervals);

    return implode(' or ', $outStrings);
  }

  /**
   * Simplifies a provided expression and validates it.
   * If the provided expression turns out to be an empty interval
   * ]0,0[ or ]0.0,0.0[ will be returned.
   * @access static
   * 
   * @param string $expression
   *   The interval to compute.
   * 
   * @throws MathIntervalException
   *   - If the provided expression is invalid.
   *   - If the lower bound of the interval is higher then its upper bound.
   * 
   * @return String
   *   The string representation of the interval.
   */
  static function compute($expression) {
    if (preg_match('/^'.MATH_INTERVAL_REGEX.'$/', $expression)) {
      $interval = new MathInterval($expression);
      return $interval->__toString();
    }
    elseif (preg_match('/^'.MATH_INTERVAL_EXPRESSION_REGEX.'$/', $expression, $results)) {
      // There are no atoms to extract.
      // Proceed to compute.
      $atom = $results[0];
      $ranges = preg_split('/ /', $atom);
      
      // First element is always an interval.
      // The regex already validated this.
      $working_interval = new MathIntervalCollection(array_shift($ranges));
      $op = current($ranges);
      
      do {
        $exp = next($ranges);
        
        if ($op == 'or') {
          $working_interval->union($exp);
        }
        elseif ($op == 'and') {
          $working_interval->intersection($exp);
        } 
      }
      while ($op = next($ranges));
      
      return $working_interval->__toString();
    }
    elseif (preg_match('/'.MATH_INTERVAL_EXP_ATOM_REGEX.'/', $expression, $atoms)) {
      // Extracted atom.
      // $atoms[0] contains the expression with parenthesis.
      // $atoms[1] contains the expression without parenthesis.
      $simplified = MathIntervalCollection::compute($atoms[1]);
      // Replace the atom with the simplified expression.
      // If multiple matches are found all will be replaced.
      $expression = str_replace($atoms[0], $simplified, $expression);
      // Run the compute again with the new expression.
      return MathIntervalCollection::compute($expression);
    }
    else {
      throw new MathIntervalException("Invalid expression.");
    }
  }

  /**
   * Check whether a given value fits within the interval.
   * @access public
   * 
   * @param int $value
   *   The value to validate
   * 
   * @return Boolean
   *   The result of the validation.
   */
  public function inInterval($value) {
    // To be true the value needs to belong in just one of the intervals.
    foreach ($this as $interval) {
      if ($interval->inInterval($value)) {
        return TRUE;
      }
    }
    return FALSE;
  }

  /**
   * Check whether the interval collection is empty.
   * @access public
   * 
   * @return Boolean
   *   Whether the interval collection is empty.
   */
  public function isEmpty() {
    // When an interval collection is empty there's only one interval
    // in the intervals array. If there were more than one empty
    // intervals they would be merged.
    return $this->totalIntervals() == 1 ? $this->get(0)->isEmpty() : FALSE;
  }

  /**
   * Check whether the interval collection allows floating values.
   * @access public
   * 
   * @return Boolean
   *   Whether the interval collection allows floating values.
   */
  public function allowFloats() {
    // If one interval allows float values all will.
    return $this->get(0)->allowFloats();
  }

  /**
   * Computes the union of this interval collection with a given one.
   * @access public
   * 
   * @param (string|MathIterval|MathIntervalCollection) $expression
   *   The interval to unite in expression form, MathInterval or
   *   MathIntervalCollection
   * 
   * @throws MathIntervalException
   *   - If the provided expression is invalid.
   *   - If the lower bound of the interval is higher then its upper bound.
   * 
   * @return MathIntervalCollection
   *   The updated interval.
   */
  public function union($expression) {
    if ($expression instanceof MathIntervalCollection) {
      $colToJoin = $expression;
    }
    else {
      if ($expression instanceof MathInterval) {
        $expression = $expression->__toString();
      }
      $colToJoin = new MathIntervalCollection($expression);
    }
    
    $intervalsToAdd = array();
    // Loop over each interval in the collection to join.
    foreach ($colToJoin as $intervalToJoin) {
      $joined = FALSE;
      // Try to join with the intervals in this collection.
      // With unions is enough to join with one.
      // If a union unites with more than one interval
      // it will be taken care when the intervals are joined between themselves.
      foreach ($this as $interval) {
        if ($interval->union($intervalToJoin)) {
          $joined = TRUE;
          break;
        }
      }
      if (!$joined) {
        // Not possible to join. Add.
        $intervalsToAdd[] = $intervalToJoin;
      }
    }
    $this->intervals = array_merge($this->intervals, $intervalsToAdd);

    // After all intervals from the other collection have been merged or added,
    // merge the existing intervals to ensure that overlapping intervals
    // are combined.
    $this->_selfUnion();
    
    return $this;
  }

  /**
   * Computes the intersection of this interval with a given one.
   * @access public
   * 
   * @param string $expression
   *   The interval to intersect.
   * 
   * @throws MathIntervalException
   *   - If the provided expression is invalid.
   *   - If the lower bound of the interval is higher then its upper bound.
   * 
   * @return MathInterval
   *   The updated interval.
   */
  public function intersection($expression) {
    if ($expression instanceof MathIntervalCollection) {
      $colToJoin = $expression;
    }
    else {
      if ($expression instanceof MathInterval) {
        $expression = $expression->__toString();
      }
      $colToJoin = new MathIntervalCollection($expression);
    }

    $newIntervals = array();
    // Loop over each interval in the collection to join.
    foreach ($colToJoin as $intervalToJoin) {
      // Each interval of the collection needs to be intersected with each
      // interval of the interval array.
      //
      // For example:
      // $c = new MathIntervalCollection('[1,10] or [16,20]');
      // $c->intersection(']7,9] or [15,18]');
      //
      // ]7,9] needs to intersect with [1,10] and with [16,20]
      // which results in ]7,9]
      // [15,18] also needs to intersect with [1,10] and with [16,20]
      // which results in [16,18]
      //
      // Merging the two results we get ]7,9] or [16,18]
      // This would have been the equivalent of doing:
      // ([1,10] or [16,20]) and (]7,9] or [15,18])
      //
      // If the intervals were updated on the go the result would be
      // an empty set.
      $copy = array();
      foreach ($this as $interval) {
        // Create a copy of the interval object before the intersection.
        $copyInterval = clone $interval;
        $copyInterval->intersection($intervalToJoin);
        $copy[] = $copyInterval;
      }

      $newIntervals = array_merge($newIntervals, $copy);
    }
    $this->intervals = $newIntervals;

    // After all intervals from the other collection have been merged or added,
    // merge the existing intervals to ensure that overlapping intervals
    // are combined.
    $this->_selfUnion();
  }

  /**
   * Recursively merges all the intervals in the array to flatten it
   * as much as possible.
   */
  private function _selfUnion() {
    while (TRUE) {
      $tot = $this->totalIntervals();
      //print($this . "\n");
      for ($i=0; $i < $tot; $i++) {
        for ($j=$i+1; $j < $tot; $j++) {
          //print('Comparing: ' . $this->get($i) . ' with ' . $this->get($j) . "\n");
          if ($this->get($i)->union($this->get($j))) {
            $this->remove($j);
            continue 3;
          }
        }
      }
      break;
    }
  }

  /**
   * Returns the interval count.
   * 
   * @return int
   *   The total number of intervals
   */
  public function totalIntervals() {
    return count($this->intervals);
  }

  /**
   * Returns interval at the given position.
   * 
   * @param int
   *   Position of the interval.
   * 
   * @return MathInterval
   *   The interval.
   */
  private function get($position) {
    return $this->intervals[$position];
  }

  /**
   * Removes interval at the given position.
   * 
   * @param int
   *   Position of the interval.
   * 
   */
  private function remove($position) {
    $interval = $this->intervals[$position];
    unset($this->intervals[$position]);
    // Re-order array.
    $this->intervals = array_values($this->intervals);
    return $interval;
  }

  /**
   * Iterator method.
   * Rewinds the array.
   */
  public function rewind() {
    $this->position = 0;
  }

  /**
   * Iterator method.
   * Returns the current value.
   */
  public function current() {
    return $this->intervals[$this->position];
  }

  /**
   * Iterator method.
   * Returns the current key.
   */
  public function key() {
    return $this->position;
  }

  /**
   * Iterator method.
   * Moves the internal pointer to the next value.
   */
  public function next() {
    ++$this->position;
  }

  /**
   * Iterator method.
   * Returns Whether the current position is valid.
   */
  public function valid() {
    return isset($this->intervals[$this->position]);
  }
}
